csi/controller.go

```diff
@@ -0,0 +1,230 @@
+	dlog.Debugf("ValidateVolumeCapabilities of id %s "+
+		"capabilities %#v "+
+		"version %#v "+
+		"attributes %#v ",
+		id,
+		capabilities,
+		version,
+		attributes)
+
+	// Check ID is valid with the specified volume capabilities
+	volumes, err := s.driver.Inspect([]string{id})
+	if err != nil || len(volumes) == 0 {
+		return nil, status.Error(codes.NotFound, "ID not found")
+	}
+	if len(volumes) != 1 {
+		errs := fmt.Sprintf(
+			"Driver returned an unexpected number of volumes when one was expected: %d",
+			len(volumes))
+		dlog.Errorln(errs)
+		return nil, status.Error(codes.Internal, errs)
+	}
+	v := volumes[0]
+	if v.Id != id {
+		errs := fmt.Sprintf(
+			"Driver volume id [%s] does not equal requested id of: %s",
+			v.Id,
+			id)
+		dlog.Errorln(errs)
+		return nil, status.Error(codes.Internal, errs)
+	}
+
+	// Setup uninitialized response object
+	result := &csi.ValidateVolumeCapabilitiesResponse_Result{
+		Supported: true,
+	}
+	resp := &csi.ValidateVolumeCapabilitiesResponse{
+		Reply: &csi.ValidateVolumeCapabilitiesResponse_Result_{
+			Result: result,
+		},
+	}
+
+	// Check capability
+	for _, capability := range capabilities {
+		// Currently the CSI spec defines all storage as "file systems."
+		// So we do not need to check this with the volume. All we will check
+		// here is the validity of the capability access type.
+		if capability.GetMount() == nil && capability.GetBlock() == nil {
+			return nil, status.Error(
+				codes.InvalidArgument,
+				"Cannot have both mount and block be undefined")
+		}
+
+		// Check access mode is setup correctly
+		mode := capability.GetAccessMode()
+		switch {
+		case mode.Mode == csi.VolumeCapability_AccessMode_SINGLE_NODE_WRITER:
+			if v.Spec.Shared {
+				result.Supported = false
+				result.Message = "Volume is a multi node volume"
+				break
+			}
+			if v.Readonly {
+				result.Supported = false
+				result.Message = "Volume is read only"
+				break
+			}
+		case mode.Mode == csi.VolumeCapability_AccessMode_SINGLE_NODE_READER_ONLY:
+			if v.Spec.Shared {
+				result.Supported = false
+				result.Message = "Volume is a multi node volume"
+				break
+			}
+			if !v.Readonly {
+				result.Supported = false
+				result.Message = "Volume is not read only"
+				break
+			}
+		case mode.Mode == csi.VolumeCapability_AccessMode_MULTI_NODE_READER_ONLY:
+			if !v.Spec.Shared {
+				result.Supported = false
+				result.Message = "Volume is not a multi node volume"
+				break
+			}
+			if !v.Readonly {
+				result.Supported = false
+				result.Message = "Volume is not read only"
+				break
+			}
+		case mode.Mode == csi.VolumeCapability_AccessMode_MULTI_NODE_SINGLE_WRITER ||
+			mode.Mode == csi.VolumeCapability_AccessMode_MULTI_NODE_MULTI_WRITER:
+			if !v.Spec.Shared {
+				result.Supported = false
+				result.Message = "Volume is not a multi node volume"
+				break
+			}
+			if v.Readonly {
+				result.Supported = false
+				result.Message = "Volume is read only"
+				break
+			}
+		default:
```

Comment:
IMO the default case should also return the result.Supported as false, instead of error.

Reply:
default is a coding error on the client and a spec contract break